oscovida/covid19dh.py

```diff
@@ -0,0 +1,341 @@
+        default `False`
+    vintage : bool, optional
+        Warning: Experimental. If `True`, downloads vintage data, by default `False`
+
+    Returns
+    -------
+    pd.DataFrame
+        DataFrame containing the covid data for a specified level, filtered by
+        the specified country. The index is set to `date`, and each row contains
+        the cumulative numbers for a region up to the specified date.
+    """
+    end = datetime.datetime.now() if end is None else end
+    end = _parse_date(end)
+    start = _parse_date(start)
+
+    if level not in {1, 2, 3}:
+        raise IndexError("invalid level, level should be 1, 2, or 3.")
+
+    if start > end:
+        raise ValueError(f"start date {start} is later than {end}")
+
+    if cache and cache._cached(level, end, raw, vintage):
+        df = cache.read(level, end, raw, vintage)
+    else:
+        df = _download(level, end, raw, vintage)
+        if cache:
+            cache.write(df, level, end, raw, vintage)
+
+    if country is not None:
+        #  We get all of the columns that give a region its name, check if the
+        #  given country argument matches any of the values in those columns
+        #  (with the country as given, or all upper case, or all lower case),
+        #  then create a mask by finding any columns that match the country
+        cols = [
+            'iso_alpha_2',
+            'iso_alpha_3',
+            'iso_numeric',
+            'administrative_area_level_1',
+        ]
+        # Some columns are missing for vintage data so we check they exist here
+        cols = list(filter(lambda x: x in df.columns, cols))
+
+        df_mask = df[cols].isin([country, country.upper(), country.lower()]).any(axis=1)
+
+        df = df[df_mask]
+
+        if len(df) == 0:
+            raise NoDataAvailable(
+                f"No data found for '{country}' in {cols}. Is '{country}' "
+                f"a valid country name or ISO country code?"
+            )
+
+    if start is not None:
+        df = df[df.index >= start]  # type: ignore
+
+    if end is not None:
+        df = df[df.index <= end]  # type: ignore
+
+    df = df.sort_values(by=["id", "date"])  # type: ignore
+
+    return df  # type: ignore
+
+
+def cite(x: pd.DataFrame, raw: bool = False) -> List[str]:
+    """Get a list of sources for a given covid19dh DataFrame
+
+    Parameters
+    ----------
+    x : pd.DataFrame
+        Input covid19dh DataFrame
+    raw : bool, optional
+        Set to `True` if the data is raw data, by default `False`
+
+    Returns
+    -------
+    List[str]
+        List of sources, each element is a string
+    """
+    # get sources
+    url = 'https://storage.covid19datahub.io/src.csv'
+    response = requests.get(url)
+    sources: pd.DataFrame = pd.read_csv(StringIO(response.text))  # type: ignore
+
+    # transform data
+    isos = set(x["iso_alpha_3"].unique())
+    params = set(x.columns)
+    # add universal
+    isos.add(math.nan)
```

Comment:
Don't have a clue. Looks like it puts `math.nan` into the `isos` set. My guess is that this is here just so that `isos` isn't an empty set, but I can't really see why it being an empty set would have any effect on what happens later on.